### PR TITLE
prefit xgboost model in `xrf_fit()`, allow early stopping

### PR DESCRIPTION
**Pull request by simonpcouch**
This PR:

* Adds support for early stopping with `xrf::xrf()`
* Corrects routing of `mtry` from `colsample_bytree` to `colsample_bynode` (and is now thus consistent with parsnip, see tidymodels/parsnip#495)
* De-duplicates xgboost training code between parsnip and rules
* Extends testing for `xrf` engine and removes `mtry` tests that are now redundant with parsnip

It does so by making use of the `prefit_xgb` argument to `xrf::xrf()`—instead of having `xrf` handle the xgboost training, we use `parsnip::xgb_train` and pass the pre-fit xgboost model to `xrf`.

<details>
  <summary> Justification for pre-fitting </summary>
The pre-fitting makes some parts of wrapping easier for us, but was initially a need for early stopping:

 `xrf::xrf.formula()` wraps both `xgboost::xgb.train()` and `glmnet::glmnet()`. `xrf::xrf.formula()` takes in an `xgb_control` argument, where all arguments passed to `xgb_control` are passed to `xgboost::xbg.train()`'s `param` argument—this is currently the only way one can pass arguments to `xgboost::xgb.train()` through `xrf::xrf.formula()`.

This is an issue for early stopping, as `early_stopping_rounds` is a _main_ argument to `xgboost::xgb.train()` and cannot be passed through `param`. Thus, if we wanted to implement an interface for early stopping, we'd either:

* need to contribute to the xrf package to allow for passing arguments to `xrf::xrf.formula()` as main arguments to `xgboost::xbg.train()`
* "prefit" the xgboost model with our own machinery and pass it as `prefit_xgb` to `xrf::xrf.formula()`.

 </details>

Related to tidymodels/parsnip#749!

### CODE REVIEW

File: tests/testthat/test-rule-fit-regression.R [hunk 1]
```diff
@@ -157,162 +157,64 @@ test_that("tidy method - regression", {
...
+    set_mode("regression")
 
-  expect_equal(
-    extract_fit_engine(pars_fit_9)$xgb$params$colsample_bytree,
-    1
+  expect_warning(
+    fit(rf_mod, mpg ~ ., data = mtcars),
+    "and were removed: xgb_control"
   )
-
```
Comment (EmilHvitfeldt) [inline tests/testthat/test-rule-fit-regression.R:312]: Why were these tests removed?

File: tests/testthat/test-rule-fit-regression.R [hunk 2]
```diff
@@ -157,162 +157,64 @@ test_that("tidy method - regression", {
...
+test_that("xrf_fit guards xgb_control", {
+  rf_mod <-
+    rule_fit(trees = 3) %>%
+    set_engine("xrf", xgb_control = list(nrounds = 3)) %>%
+    set_mode("regression")
 
-  expect_equal(
-    extract_fit_engine(pars_fit_9)$xgb$params$colsample_bytree,
-    1
+  expect_warning(
```
Comment (EmilHvitfeldt) [inline tests/testthat/test-rule-fit-regression.R:216]: This should use `expect_snapshot()`

File: tests/testthat/test-rule-fit-regression.R [hunk 3]
```diff
@@ -157,162 +157,64 @@ test_that("tidy method - regression", {
...
-        Sale_Price ~ Neighborhood + Longitude + Latitude +
-          Gr_Liv_Area + Central_Air,
-        data = ames_data$ames
-      )
-  })
-
-  expect_equal(
-    extract_fit_engine(pars_fit_3)$xgb$params$colsample_bytree,
-    1
+  expect_warning(
```
Comment (EmilHvitfeldt) [inline tests/testthat/test-rule-fit-regression.R:184]: This should use `expect_snapshot()`

**Review by EmilHvitfeldt — CHANGES_REQUESTED** [review]
Couple of small changes to the tests and I think we are good here!

File: tests/testthat/test-rule-fit-regression.R (hunk 2 above)
Comment (simonpcouch) [reply tests/testthat/test-rule-fit-regression.R:216]: Agreed! Thanks for these two nudges.

I'm seeing:

```
── Error (test-rule-fit-regression.R:184:3): early stopping works in xrf_fit ───
Error in `!cran`: invalid argument type
Backtrace:
 1. testthat::expect_snapshot(...)
      at test-rule-fit-regression.R:184:2
 2. testthat:::expect_snapshot_helper(...)

── Error (test-rule-fit-regression.R:216:3): xrf_fit guards xgb_control ────────
Error in `!cran`: invalid argument type
Backtrace:
 1. testthat::expect_snapshot(...)
      at test-rule-fit-regression.R:216:2
 2. testthat:::expect_snapshot_helper(...)

[ FAIL 2 | WARN 0 | SKIP 0 | PASS 30 ]
```

when I try to generate these snapshots. Can you confirm that you also see this? I've generally ran into a few oddities with the testing environment with this package—I'm also unable to run the `make_*` data helpers interactively.

Comment (EmilHvitfeldt) [reply tests/testthat/test-rule-fit-regression.R:216]: I'm not able to reproduce this issue :( It works flawlessly for me 🤞

**Comment by simonpcouch** [comment]
Ah, figured it out🙈 Wrapping in `suppressMessages` pending https://github.com/holub008/xrf/pull/21 merged + on CRAN.

**Review by EmilHvitfeldt — APPROVED** [review]

**Pull request merged**